File: kzen-auto-jvm/src/main/java/tech/kzen/auto/server/objects/report/pipeline/input/model/RecordTextFlyweight.java
package tech.kzen.auto.server.objects.report.pipeline.input.model;


import org.jetbrains.annotations.NotNull;

import java.util.Arrays;


public class RecordTextFlyweight
        implements CharSequence
{
    //-----------------------------------------------------------------------------------------------------------------
    public static final RecordTextFlyweight empty = standalone("");


    public static final long[] decimalLongPowers = {
            1,
            10,
            100,
            1_000,
            10_000,
            100_000,
            1_000_000,
            10_000_000,
            100_000_000,
            1_000_000_000,
            10_000_000_000L,
            100_000_000_000L,
            1_000_000_000_000L,
            10_000_000_000_000L,
            100_000_000_000_000L,
            1_000_000_000_000_000L,
            10_000_000_000_000_000L,
            100_000_000_000_000_000L
    };


    public static final int maxLongDecimalLength = Long.valueOf(999999999999999999L).toString().length();


    public static RecordTextFlyweight standalone(String value) {
        char[] chars = value.toCharArray();
        RecordItemBuffer buffer = RecordItemBuffer.ofSingle(chars, 0, chars.length);
        RecordTextFlyweight flyweight = new RecordTextFlyweight();
        flyweight.selectHostValue(buffer,0, chars.length);
        return flyweight;
    }


    //-----------------------------------------------------------------------------------------------------------------
    private RecordItemBuffer host;
//    private int fieldIndex = -1;

    private int valueOffset = -1;
    private int valueLength = -1;

    private int hashCodeCache = -1;


    //-----------------------------------------------------------------------------------------------------------------
    public void selectHost(RecordItemBuffer host) {
        this.host = host;
        valueOffset = -1;
        valueLength = -1;
    }


    public void selectHostValue(RecordItemBuffer host, int valueOffset, int valueLength) {
        this.host = host;
        selectValue(valueOffset, valueLength);
    }


    public void selectHostField(RecordItemBuffer host, int fieldIndex) {
        this.host = host;
        selectField(fieldIndex);
    }


    public void selectField(int fieldIndex) {
        int startIndex;
        int endIndex;
        if (fieldIndex == 0) {
            startIndex = 0;
            endIndex = host.fieldEnds[0];
        }
        else {
            startIndex = host.fieldEnds[fieldIndex - 1];
            endIndex = host.fieldEnds[fieldIndex];
        }

        int length = endIndex - startIndex;

        selectValue(startIndex, length);
    }


    public void selectValue(int valueOffset, int valueLength) {
        this.valueOffset = valueOffset;
        this.valueLength = valueLength;
        hashCodeCache = -1;
    }


    //-----------------------------------------------------------------------------------------------------------------
    @Override
    public int length() {
        return valueLength;
    }


    @Override
    public char charAt(int index) {
        return host.fieldContents[valueOffset + index];
    }


    @Override
    public CharSequence subSequence(int start, int end) {
        throw new UnsupportedOperationException();
    }


    //-----------------------------------------------------------------------------------------------------------------
    public RecordTextFlyweight detach() {
        RecordItemBuffer buffer = RecordItemBuffer.ofSingle(host.fieldContents, valueOffset, valueLength);
        RecordTextFlyweight detached = new RecordTextFlyweight();
        detached.host = buffer;
//        detached.fieldIndex = 0;
        detached.valueOffset = 0;
        detached.valueLength = valueLength;
        detached.hashCodeCache = hashCodeCache;
        return detached;
    }


    //-----------------------------------------------------------------------------------------------------------------
    public void trim() {
        char[] contents = host.fieldContents;

        int initialLength = valueLength;
        for (int i = 0; i < initialLength; i++) {
            if (! Character.isWhitespace(contents[valueOffset])) {
                break;
            }

            valueOffset++;
            valueLength--;
        }

        int afterLeftTrimLength = valueLength;
        for (int i = (afterLeftTrimLength - 1); i >= 0; i--) {
            if (! Character.isWhitespace(contents[valueOffset + i])) {
                break;
            }

            valueLength--;
        }
    }


    //-----------------------------------------------------------------------------------------------------------------
    public boolean isDecimal() {
        char[] contents = host.fieldContents;
        int len = valueLength;
        int offset = valueOffset;

        var dotCount = 0;
        var digitCount = 0;

        var i = 0;
        while (i < len) {
            char nextChar = contents[offset + i++];

            if (nextChar == '.') {
                dotCount++;
            }
            else if ('0' <= nextChar && nextChar <= '9') {
                digitCount++;
            }
            else {
                return false;
            }
        }

        return digitCount > 0 && dotCount <= 1;
    }


    public long toLong() {
        return toLong(0, valueLength);
    }

    // Shadowed from Java Util Lang, but specialized
    public long toLong(int offset, int length) {
        if (offset + length > valueLength) {
            throw new NumberFormatException(
                    offset + " - " + length + " - " + valueOffset + " - " + valueLength + " - " + toString());
        }
        if (length <= 0) {
            throw new NumberFormatException("offset = " + offset+ ", length = " + length + ": " + toString());
        }

        char[] contents = host.fieldContents;
        int relativeOffset = valueOffset + offset;

        boolean negative = false;
        int i = 0;
        long limit = -Long.MAX_VALUE;
        int digit;

        char firstChar = contents[relativeOffset + i];
        if (firstChar < '0') { // Possible leading "+" or "-"
            if (firstChar == '-') {
                negative = true;
                limit = Long.MIN_VALUE;
            }
            else if (firstChar != '+') {
                throw new NumberFormatException(toString());
            }
            i++;
            if (length == 1) { // Cannot have lone "+" or "-"
                throw new NumberFormatException(toString());
            }
        }

        long multmin = limit / 10;
        var result = 0L;
        while (i < length) {
            // Accumulating negatively avoids surprises near MAX_VALUE
            digit = contents[relativeOffset + i++] - '0';
            if (digit < 0 || digit > 9) {
                throw new NumberFormatException(toString());
            }
            if (result < multmin) {
                throw new NumberFormatException(toString());
            }
            result *= 10;
            if (result < limit + digit) {
                throw new NumberFormatException(toString());
            }
            result -= digit;
        }

        return negative ? result : -result;
    }


    public double toDouble() {
        char[] contents = host.fieldContents;
        int len = valueLength;
        int offset = valueOffset;

        int pointIndex = -1;
        for (int i = 0; i < len; i++) {
            if (contents[offset + i] == '.') {
                pointIndex = i;
                break;
            }
        }

        if (pointIndex == -1) {
            return (double) toLong();
        }
        else if (pointIndex == len - 1) {
            return (double) toLong(0, len - 1);
        }

        long wholePart =
                pointIndex == 0  || pointIndex == 1 && (contents[offset] == '+' || contents[offset] == '-')
                ? 0
                : Math.abs(toLong(0, pointIndex));

        // https://math.stackexchange.com/questions/64042/what-are-the-numbers-before-and-after-the-decimal-point-referred-to-in-mathemati/438718#438718
        int fractionDigits = len - pointIndex - 1;
        int fractionLeadingZeroes = 0;
        for (int i = 1; i <= fractionDigits; i++) {
            if (contents[offset + pointIndex + i] != '0') {
                break;
            }
            fractionLeadingZeroes++;
        }

        int fractionDigitsWithoutLeadingZeroes = fractionDigits - fractionLeadingZeroes;
        long fractionAsLongWithoutLeadingZeroes =
                fractionDigitsWithoutLeadingZeroes == 0
                ? 0
                : toLong(pointIndex + fractionLeadingZeroes + 1, fractionDigitsWithoutLeadingZeroes);

        double fractionalPartWithoutLeadingZeroes =
                (double) fractionAsLongWithoutLeadingZeroes / decimalLongPowers[fractionDigitsWithoutLeadingZeroes];

        double fractionalPart =
                fractionLeadingZeroes == 0
                ? fractionalPartWithoutLeadingZeroes
                : fractionalPartWithoutLeadingZeroes / decimalLongPowers[fractionLeadingZeroes];

        double value = wholePart + fractionalPart;
        return contents[offset] == '-' ? -value : value;
    }


    public double toDoubleOrNan() {
        int len = valueLength;
        if (len == 0) {
            return Double.NaN;
        }

        char[] contents = host.fieldContents;
        int offset = valueOffset;

        int leadingZeroes = 0;
        int pointIndex = -1;
        for (int i = 0; i < len; i++) {
            char nextChar = contents[offset + i];
            if (nextChar == '.') {
                if (pointIndex != -1 ||
                        len == 1 ||
                        len == 2 && (contents[offset] == '+' || contents[offset] == '-')
                ) {
                    return Double.NaN;
                }
                else {
                    pointIndex = i;
                }
            }
            else if (nextChar == '+' || nextChar == '-') {
                if (i != 0 || len == 1) {
                    return Double.NaN;
                }
            }
            else if (! ('0' <= nextChar && nextChar <= '9')) {
                return Double.NaN;
            }
            else if (nextChar == '0' && leadingZeroes == i) {
                leadingZeroes++;
            }
        }

        if (pointIndex == -1) {
            if (valueLength - leadingZeroes > maxLongDecimalLength) {
                return Double.NaN;
            }
            return (double) toLong();
        }
        else if (pointIndex == len - 1) {
            if (valueLength - leadingZeroes - 1 > maxLongDecimalLength) {
                return Double.NaN;
            }
            return (double) toLong(0, len - 1);
        }

        long wholePart;
        if (pointIndex == 0 || pointIndex == 1 && (contents[offset] == '+' || contents[offset] == '-')) {
            wholePart = 0;
        }
        else if (pointIndex - leadingZeroes > maxLongDecimalLength) {
            return Double.NaN;
        }
        else {
            wholePart = Math.abs(toLong(0, pointIndex));
        }

        int fractionDigits = len - pointIndex - 1;
        int fractionLeadingZeroes = 0;
        for (int i = 1; i <= fractionDigits; i++) {
            if (contents[offset + pointIndex + i] != '0') {
                break;
            }
            fractionLeadingZeroes++;
        }

        int fractionDigitsWithoutLeadingZeroes = fractionDigits - fractionLeadingZeroes;
        long fractionAsLongWithoutLeadingZeroes;
        if (fractionDigitsWithoutLeadingZeroes == 0) {
            fractionAsLongWithoutLeadingZeroes = 0;
        }
        else if (fractionDigitsWithoutLeadingZeroes > maxLongDecimalLength) {
            return Double.NaN;
        }
        else {
            fractionAsLongWithoutLeadingZeroes = toLong(
                    pointIndex + fractionLeadingZeroes + 1, fractionDigitsWithoutLeadingZeroes);
        }

        double fractionalPartWithoutLeadingZeroes =
                (double) fractionAsLongWithoutLeadingZeroes / decimalLongPowers[fractionDigitsWithoutLeadingZeroes];

        double fractionalPart =
                fractionLeadingZeroes == 0
                ? fractionalPartWithoutLeadingZeroes
                : fractionalPartWithoutLeadingZeroes / decimalLongPowers[fractionLeadingZeroes];

        double value = wholePart + fractionalPart;
        return contents[offset] == '-' ? -value : value;
    }


    //-----------------------------------------------------------------------------------------------------------------
    @NotNull
    @Override
    public String toString() {
        return new String(host.fieldContents, valueOffset, valueLength);
    }


    @Override
    public int hashCode() {
        if (hashCodeCache != -1) {
            return hashCodeCache;
        }

        char[] contents = host.fieldContents;
        int offset = valueOffset;
        int end = offset + valueLength;

        var result = 1;
        for (int i = offset; i < end; i++) {
            result = 31 * result + contents[i];
        }

        hashCodeCache = result;
        return result;
    }


    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }

        if (other.getClass() != RecordTextFlyweight.class) {
            return false;
        }

        RecordTextFlyweight that = (RecordTextFlyweight) other;

        int len = valueLength;
        if (len != that.valueLength) {
            return false;
        }

        if (hashCode() != that.hashCode()) {
            return false;
        }

        char[] contents = host.fieldContents;
        int offset = valueOffset;
        char[] thatContents = that.host.fieldContents;
        int thatValueOffset = that.valueOffset;

        return Arrays.equals(
                contents, offset, offset + len,
                thatContents, thatValueOffset, thatValueOffset + len);
    }
}

File: kzen-auto-jvm/src/main/kotlin/tech/kzen/auto/server/objects/report/input/model/RecordTextFlyweight.kt
package tech.kzen.auto.server.objects.report.input.model

import java.util.*


class RecordTextFlyweight(
    val recordLineBuffer: RecordLineBuffer,
):
    CharSequence
{
    //-----------------------------------------------------------------------------------------------------------------
    companion object {
        val empty = standalone("")

        private val decimalLongPowers = longArrayOf(
            1,
            10,
            100,
            1_000,
            10_000,
            100_000,
            1_000_000,
            10_000_000,
            100_000_000,
            1_000_000_000,
            10_000_000_000,
            100_000_000_000,
            1_000_000_000_000,
            10_000_000_000_000,
            100_000_000_000_000,
            1_000_000_000_000_000,
            10_000_000_000_000_000,
            100_000_000_000_000_000,
            1_000_000_000_000_000_000
        )

        fun standalone(value: String): RecordTextFlyweight {
            val chars = value.toCharArray()
            val buffer = RecordLineBuffer.of(chars, 0, chars.size)
            buffer.selectFlyweight(0)
            return buffer.flyweight
        }

//        fun standaloneSlow(value: String): RecordTextFlyweight {
//            val chars = value.toCharArray()
//            val buffer = RecordLineBuffer()
//            for (c in chars) {
//                buffer.addToField(c)
//            }
//            buffer.commitField()
//            buffer.selectFlyweight(0)
//            return buffer.flyweight
//        }
    }


    //-----------------------------------------------------------------------------------------------------------------
    private var fieldIndex = -1
    private var valueOffset = 0
    private var valueLength = 0

    private var hashCodeCache: Int = -1


    //-----------------------------------------------------------------------------------------------------------------
    fun select(fieldIndex: Int, valueOffset: Int, valueLength: Int) {
        this.fieldIndex = fieldIndex
        this.valueOffset = valueOffset
        this.valueLength = valueLength
        hashCodeCache = -1
    }


    //-----------------------------------------------------------------------------------------------------------------
    override val length: Int
        get() = valueLength


    override fun get(index: Int): Char {
        return recordLineBuffer.fieldContents[valueOffset + index]
    }


    override fun subSequence(startIndex: Int, endIndex: Int): CharSequence {
        TODO("Not yet implemented")
    }


    //-----------------------------------------------------------------------------------------------------------------
    fun detach(): RecordTextFlyweight {
        val buffer = RecordLineBuffer.of(recordLineBuffer.fieldContents, valueOffset, valueLength)
        val detached = buffer.flyweight
        detached.fieldIndex = 0
        detached.valueOffset = 0
        detached.valueLength = valueLength
        detached.hashCodeCache = hashCodeCache
        return detached
    }


    //-----------------------------------------------------------------------------------------------------------------
    fun trim() {
        val contents = recordLineBuffer.fieldContents

        val initialLength = valueLength
        for (i in 0 until initialLength) {
            if (! Character.isWhitespace(contents[valueOffset])) {
                break
            }

            valueOffset++
            valueLength--
        }

        val afterLeftTrimLength = valueLength
        for (i in (afterLeftTrimLength - 1) downTo 0) {
            if (! Character.isWhitespace(contents[valueOffset + i])) {
                break
            }

            valueLength--
        }
    }


    //-----------------------------------------------------------------------------------------------------------------
//    @Suppress("ConvertTwoComparisonsToRangeCheck")
    fun isDecimal(): Boolean {
        val contents = recordLineBuffer.fieldContents
        val len = valueLength
        val offset = valueOffset

        var dotCount = 0
        var digitCount = 0

        var i = 0
        while (i < len) {
            when (contents[offset + i++]) {
                '.' ->
                    dotCount++

                in '0'..'9' ->
                    digitCount++

                else ->
                    return false
            }
        }

        return digitCount > 0 && dotCount <= 1
    }


    // Shadowed from Java Util Lang, but specialized
    fun toLong(
        offset: Int = 0,
        length: Int = valueLength
    ): Long {
        if (offset + length > valueLength) {
            throw NumberFormatException("$offset - $length - $valueOffset - $valueLength - " + toString())
        }
        if (length <= 0) {
            throw NumberFormatException(toString())
        }

        val contents = recordLineBuffer.fieldContents
        val relativeOffset = valueOffset + offset

        var negative = false
        var i = 0
        var limit = -Long.MAX_VALUE
        var digit: Int

        val firstChar = contents[relativeOffset + i]
        if (firstChar < '0') { // Possible leading "+" or "-"
            if (firstChar == '-') {
                negative = true
                limit = Long.MIN_VALUE
            }
            else if (firstChar != '+') {
                throw NumberFormatException(toString())
            }
            i++
            if (length == 1) { // Cannot have lone "+" or "-"
                throw NumberFormatException(toString())
            }
        }

        val multmin = limit / 10
        var result = 0L
        while (i < length) {
            // Accumulating negatively avoids surprises near MAX_VALUE
            digit = contents[relativeOffset + i++] - '0'
            if (digit < 0 || digit > 9) {
                throw NumberFormatException(toString())
            }
            if (result < multmin) {
                throw NumberFormatException(toString())
            }
            result *= 10
            if (result < limit + digit) {
                throw NumberFormatException(toString())
            }
            result -= digit
        }

        return if (negative) result else -result
    }


    fun toDouble(): Double {
        val contents = recordLineBuffer.fieldContents
        val len = valueLength
        val offset = valueOffset

        var pointIndex = -1
        for (i in 0 until len) {
            if (contents[offset + i] == '.') {
                pointIndex = i
                break
            }
        }

        if (pointIndex == -1) {
            return toLong().toDouble()
        }
        else if (pointIndex == len - 1) {
            return toLong(0, len - 1).toDouble()
        }

        val wholePart =
            if (pointIndex == 0) {
                0
            }
            else {
                toLong(0, pointIndex)
            }

        // https://math.stackexchange.com/questions/64042/what-are-the-numbers-before-and-after-the-decimal-point-referred-to-in-mathemati/438718#438718
        val fractionDigits = len - pointIndex - 1
        val fractionAsLong: Long = toLong(pointIndex + 1, fractionDigits)
        val factionalPart = fractionAsLong.toDouble() / decimalLongPowers[fractionDigits]
        return wholePart + factionalPart
    }


    fun toDoubleOrNan(): Double {
        val len = valueLength
        if (len == 0) {
            return Double.NaN
        }

        val contents = recordLineBuffer.fieldContents
        val offset = valueOffset

        var pointIndex = -1
        for (i in 0 until len) {
            val nextChar = contents[offset + i]
            if (nextChar == '.') {
                if (pointIndex != -1 ||
                        len == 1 ||
                        len == 2 && (
                            contents[offset] == '+' ||
                            contents[offset] == '-')
                ) {
                    return Double.NaN
                }
                else {
                    pointIndex = i
                }
            }
            else if (nextChar == '+' || nextChar == '-') {
                if (i != 0 || len == 1) {
                    return Double.NaN
                }
            }
            else if (nextChar !in '0'..'9') {
                return Double.NaN
            }
        }

        if (pointIndex == -1) {
            return toLong().toDouble()
        }
        else if (pointIndex == len - 1) {
            return toLong(0, len - 1).toDouble()
        }

        val wholePart =
            if (pointIndex == 0) {
                0
            }
            else {
                toLong(0, pointIndex)
            }

        // https://math.stackexchange.com/questions/64042/what-are-the-numbers-before-and-after-the-decimal-point-referred-to-in-mathemati/438718#438718
        val fractionDigits = len - pointIndex - 1

        var fractionLeadingZeroes = 0
        for (i in 1 .. fractionDigits) {
            if (contents[pointIndex + i] != '0') {
                break
            }
            fractionLeadingZeroes++
        }

        val factionDigitsWithoutLeadingZeroes = fractionDigits - fractionLeadingZeroes
        val fractionAsLongWithoutLeadingZeroes: Long =
            toLong(pointIndex + fractionLeadingZeroes + 1, factionDigitsWithoutLeadingZeroes)
        val factionalPartWithoutLeadingZeroes =
            fractionAsLongWithoutLeadingZeroes.toDouble() / decimalLongPowers[factionDigitsWithoutLeadingZeroes]

        val factionalPart =
            if (fractionLeadingZeroes == 0) {
                factionalPartWithoutLeadingZeroes
            }
            else {
                factionalPartWithoutLeadingZeroes / decimalLongPowers[fractionLeadingZeroes]
            }

        return wholePart + factionalPart
    }


    //-----------------------------------------------------------------------------------------------------------------
    override fun toString(): String {
        return String(recordLineBuffer.fieldContents, valueOffset, length)
    }


    override fun hashCode(): Int {
        if (hashCodeCache != -1) {
            return hashCodeCache
        }

        val contents = recordLineBuffer.fieldContents
        val offset = valueOffset
        val end = offset + valueLength

        var result = 1
        for (i in offset until end) {
            result = 31 * result + contents[i].toInt()
        }

        hashCodeCache = result
        return result
    }


    override fun equals(other: Any?): Boolean {
        val that = other as? RecordTextFlyweight
            ?: return false

        val len = valueLength
        if (len != that.valueLength) {
            return false
        }

        if (hashCode() != that.hashCode()) {
            return false
        }

        val contents = recordLineBuffer.fieldContents
        val offset = valueOffset
        val thatContents = that.recordLineBuffer.fieldContents
        val thatValueOffset = that.valueOffset

        @Suppress("ReplaceJavaStaticMethodWithKotlinAnalog")
        return Arrays.equals(
            contents, offset, offset + len,
            thatContents, thatValueOffset, thatValueOffset + len)
    }
}